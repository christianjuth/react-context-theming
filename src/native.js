import { StyleSheet } from 'react-native';
import { useTheme } from './context';
import memoize from 'fast-memoize';
export const makeStyleCreator = (fn) => {
    return memoize((props, extraData) => (StyleSheet.create(fn(props, ...extraData))));
};
export const useStyleCreator = (styleFn, ...extraData) => {
    return styleFn(useTheme(), extraData);
};
