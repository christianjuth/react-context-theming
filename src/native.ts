import { TextStyle, ViewStyle, StyleSheet } from 'react-native';
import { useTheme } from './context';
import { theme as themeType } from './types';
import memoize from 'fast-memoize';

interface StyleSheet {
  [key: string]: TextStyle | ViewStyle
}

type styleCreator = (props: any, ...extraData: any[]) => StyleSheet;

export const makeStyleCreator = (fn: (theme: themeType, ...extraData: any[]) => StyleSheet): styleCreator => {
  return memoize((props: any, extraData) => (
    StyleSheet.create(fn(props, ...extraData))
  ));
}

export const useStyleCreator = (styleFn: styleCreator, ...extraData: any[]) => {
  return styleFn(useTheme(), extraData);
};