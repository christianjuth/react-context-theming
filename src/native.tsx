import * as React from 'react';
import { TextStyle, ViewStyle, ImageStyle, StyleSheet, StyleProp } from 'react-native';
import { useTheme, Context, Theme } from './index';

export interface StyleSheet {
  [key: string]: StyleProp<TextStyle> | StyleProp<ViewStyle> | StyleProp<ImageStyle>;
}
export interface GeneratedStyles {
  [key: string]: StyleProp<any>;
}

export type StyleCreatorFunction<T> = (theme: T, ...extraData: any[]) => StyleSheet;

export interface WithStyleCreator<T = Theme> {
  theme: T,
  styles: GeneratedStyles
};



export function makeStyleCreator<T = Theme>(
  fn: StyleCreatorFunction<T>
): StyleCreatorFunction<T> {
  return (props: any, ...extraData: any[]) => (
    StyleSheet.create(fn(props, ...extraData) as any)
  );
}

export function useStyleCreator<T = Theme>(styleFn: StyleCreatorFunction<T>, ...extraData: any[]): GeneratedStyles {
  return styleFn(useTheme<T>(), ...extraData);
};

export function withStyleCreator<T = Theme>(
  Component: any, 
  styleFn: StyleCreatorFunction<T>, 
  ...extraData: any[]
): any {
  return class WrappedComponent extends React.Component<{}, null> {
    static contextType = Context;
    render() {
      return (
        <Component 
          {...this.props} 
          styles={styleFn(this.context, extraData)}
          theme={this.context}
        />
      );
    }
  };
}