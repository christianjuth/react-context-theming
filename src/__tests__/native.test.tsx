import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { View } from 'react-native';
import { Provider, defaultTheme, Theme } from '../index';
import { makeStyleCreator, useStyleCreator, withStyleCreator, WithStyleCreator } from '../native';

describe('native', () => {

  it('makeStyleCreator', () => {
    const styleCreator = makeStyleCreator(theme => ({
      container: {
        flex: 1,
        backgroundColor: theme.colors.background
      }
    }));
    expect(styleCreator(defaultTheme)).toMatchSnapshot();
  });

  it('useStyleCreator', () => {
    const styleCreator = makeStyleCreator(theme => ({
      foo: {
        height: 100,
        width: 100,
        backgroundColor: theme.colors.primary
      }
    }));

    function Foo() {
      const styles = useStyleCreator(styleCreator);
      return <View style={styles.foo}/>;
    }

    const result = renderer.create(
      <Provider theme={defaultTheme}>
        <Foo/>
      </Provider>
    ).toJSON();

    expect(result).toMatchSnapshot();
  });

  it('withStyleCreator', () => {
    const styleCreator = makeStyleCreator(theme => ({
      foo: {
        height: 100,
        width: 100,
        backgroundColor: theme.colors.primary
      }
    }));

    let theme;
    class Foo extends React.Component<WithStyleCreator> {
      render() {
        theme = this.props.theme;
        return <View style={this.props.styles.foo}/>;
      }
    }
    const FooWithStyleCreator = withStyleCreator(Foo, styleCreator);

    const result = renderer.create(
      <Provider theme={defaultTheme}>
        <FooWithStyleCreator/>
      </Provider>
    ).toJSON();

    expect(result).toMatchSnapshot();
    expect(theme).toBe(defaultTheme);
  });

});