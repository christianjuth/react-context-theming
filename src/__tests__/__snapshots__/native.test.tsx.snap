// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`native makeStyleCreator 1`] = `
Object {
  "container": Object {
    "backgroundColor": "#f5f5f5",
    "flex": 1,
  },
}
`;

exports[`native useStyleCreator 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#00bcd4",
      "height": 100,
      "width": 100,
    }
  }
/>
`;

exports[`native withStyleCreator 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#00bcd4",
      "height": 100,
      "width": 100,
    }
  }
/>
`;
